odoo/addons/base/models/ir_ui_view.py

```diff
@@ -2826,6 +2826,23 @@ def process(node, info, prefix):
         process(etree.fromstring(view_info['arch']), view_info, '')
         return result
 
+    def _ensure_xml_id(self):
+        # Prevent triggering unique constraint when updating a module where master data was added
+        # both in the code and by a user.
+        # i.e. Country state was added both in the code and in master data => shouldn't trigger
+        self.env['ir.model.data']._update_xmlids([
+            {
+                'xml_id': record._xml_id_builder(),
+                'record': record,
+                'noupdate': True,
+            }
+            for record in (
+                self.browse(meta['id'])
+                for meta in self.search([]).get_metadata()
+                if not meta.get('xmlid')
+            )
+        ])
```

Comment:
I would actually rewrite it this way for clarity (and to avoid calling the builder twice)
```suggestion
        to_generate_xids = []
        for rec in self.search([]):
            xid = rec._xml_id_builder()
            if not [
                meta for meta in rec.get_metadata()
                if meta.get("xmlid") == xid
            ]:
                to_generate_xids.append(
                    {
                        'xml_id': xid,
                        'record': rec,
                        'noupdate': True,
                    }
                )
        self.env['ir.model.data']._update_xmlids(to_generate_xids)

```

Reply:
You are actually doing one `get_metadata` per record in your version, which does one `search_read`, which is actually a lot worse than doing 2 `_xml_id_builder`.
Fixed it another way.